src/main/java/io/curity/identityserver/plugin/aws/authentication/CallbackRequestHandler.java

```diff
@@ -38,118 +28,241 @@
-    public Optional<AuthenticationResult> post(CallbackGetRequestModel requestModel,
-                                               Response response) {
-        throw _exceptionFactory.methodNotAllowed();
-    }
+    private Map<String, Object> redeemCodeForTokens(CallbackGetRequestModel requestModel)
+    {
+        HttpResponse tokenResponse = getWebServiceClient()
+                .withPath("/oauth2/token")
+                .request()
+                .contentType("application/x-www-form-urlencoded")
+                .body(getFormEncodedBodyFrom(createPostData(_config.getClientId(), _config.getClientSecret(),
+                        requestModel.getCode(), requestModel.getRequestUrl())))
+                .header("Authorization", "Basic " +
+                        Base64.getEncoder().encodeToString((_config.getClientId() + ":" +
+                                _config.getClientSecret()).getBytes()))
+                .method("POST")
+                .response();
+        int statusCode = tokenResponse.statusCode();
 
-    public Map<String, Object> getTokens(String tokenEndpoint, String clientId, String clientSecret, String code, String state) {
+        if (statusCode != 200)
+        {
+            if (_logger.isInfoEnabled())
+            {
+                _logger.info("Got error response from token endpoint: error = {}, {}", statusCode,
+                        tokenResponse.body(HttpResponse.asString()));
+            }
 
-        _oauthClient.validateState(state);
+            throw _exceptionFactory.internalServerException(ErrorCode.EXTERNAL_SERVICE_ERROR);
+        }
 
-        try {
-            ParamBuilder postData = new ParamBuilder();
-            postData.addPair(PARAM_CODE, code)
-                    .addPair(PARAM_GRANT_TYPE, PARAM_GRANT_TYPE_AUTHORIZATION_CODE)
-                    .addPair(SCOPE, AWSAuthenticatorRequestHandler.getScope(_config))
-                    .addPair(PARAM_REDIRECT_URI, _oauthClient.getCallbackUrl());
-            UrlEncodedFormEntity data = new UrlEncodedFormEntity(postData.getPairs());
+        return _json.fromJson(tokenResponse.body(HttpResponse.asString()));
+    }
 
-            HttpPost post = new HttpPost(tokenEndpoint);
-            post.addHeader(HttpHeaders.CONTENT_TYPE, CONTENT_TYPE);
-            post.setHeader(HttpHeaders.AUTHORIZATION, "Basic " + Base64.getEncoder().encodeToString((clientId + ":" + clientSecret).getBytes()));
-            post.setEntity(data);
+    private WebServiceClient getWebServiceClient()
+    {
+        Optional<HttpClient> httpClient = _config.getHttpClient();
 
-            HttpResponse response = _client.execute(post);
-            if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK) {
-                _logger.debug("Got error response from token endpoint {}", response.getStatusLine());
+        if (httpClient.isPresent())
+        {
+            return _webServiceClientFactory.create(httpClient.get()).withHost(_config.getDomain().getHost());
```

Comment:
if the HTTP client is configured with one scheme (e.g., `http`) and the domain has another (e.g., `https`), I'd log a warning.

Reply:
okay